src/MergeContext.js

```diff
@@ -454,6 +454,265 @@ class BranchPosition
+
+    // zero or more paragraphs after the parsed fields and field terminator
+    remaining() { return this._lines.join('\n'); }
+}
+
+// computes future commit message from raw PR
+class CommitMessage
+{
+    constructor(rawPr, defaultAuthor) {
+
+        // the (required) commit message title
+        this._title = rawPr.title + ' (#' + rawPr.number + ')';
+
+        // PR description has three optional parts: [header]+[body]+[trailer]
+        // The parts are separated by empty lines.
+
+        // The header is dedicated to Anubis-recognized/consumed PR metadata.
+        // The commit message only retains the body and the trailer.
+
+        // PR description without the header and trailer parts
+        this._body = null;
+
+        // optional commit attributes recognized by GitHub, git, and/or Project
+        // these attributes are partially validated by Anubis
+        // stored in the original text format
+        this._trailer = null;
+
+        // GitHub-suggested commit author
+        // uses {name, email, date} format
+        assert(defaultAuthor);
+        this._defaultAuthor = defaultAuthor;
+
+        // optional commit author specified by 'Authored-by' header field
+        // uses {name, email} format
+        // overwrites this._defaultAuthor's name and email
+        this._customAuthor = null;
+
+        this._checkRaw(this._title);
+
+        if (rawPr.body !== undefined && rawPr.body !== null)
+            this._parse(rawPr.body);
+    }
+
+    // complete message for the future commit
+    whole() {
+        assert(this._title.length); // the only required part
+        let message = this._title;
+        if (this._body !== null)
+            message += '\n\n' + this._body;
+        if (this._trailer !== null)
+            message += '\n\n' + this._trailer;
+        return message;
+    }
+
+    // the future commit author in the {name, email, date} format
+    author()
+    {
+        if (this._customAuthor === null)
+            return this._defaultAuthor;
+        return {name: this._customAuthor.name, email: this._customAuthor.email, date: this._defaultAuthor.date};
+    }
+
+    // returns the position of the first non-ASCII_printable character (or -1)
+    _invalidCharacterPosition(str) {
+        const prohibitedCharacters = /[^\u{20}-\u{7e}]/u;
+        const match = prohibitedCharacters.exec(str);
+        return match ? match.index : -1;
+    }
+
+    // removes leading empty lines and trims the end
+    _trim(str) {
+        // cannot just use trim() to preserve the first line indentation (if any).
+        return str.replace(/^\s*\n+/g, '').trimEnd();
+    }
+
+    // returns either a string containing canonical spelling of the first paragraph word
+    // that is followed by a colon, or null
+    _paragraphLabel(str) {
+        if (/^\s{4}/.test(str)) // not a regular paragraph but a quotation
+            return null;
+        const rawLabel = str.match(/^\s*(\S+)\s*:/);
```

Comment:
IIRC, we have discussed tightening this up to ban whitespace before colon:

```suggestion
        const rawLabel = str.match(/^\s*(\S+):/);
```

Reply:
Done (97ab1f8).